staking/apr/compute.go

```diff
@@ -55,152 +56,81 @@ func expectedRewardPerYear(
-	// )
-
-	// if err != nil {
-	// 	return &zero, nil
-	// }
-
-	// oneSnapshotAgo, err := bc.ReadValidatorSnapshotAtEpoch(
-	// 	oneEpochAgo,
-	// 	validatorNow.Address,
-	// )
-
-	// if err != nil {
-	// 	return &zero, nil
-	// }
-
-	// blockNumAtTwoEpochAgo, blockNumAtOneEpochAgo :=
-	// 	shard.Schedule.EpochLastBlock(twoEpochAgo.Uint64()),
-	// 	shard.Schedule.EpochLastBlock(oneEpochAgo.Uint64())
-
-	// headerOneEpochAgo, headerTwoEpochAgo, err := pastTwoEpochHeaders(bc)
-
-	// // TODO Figure out why this is happening
-	// if headerOneEpochAgo == nil || headerTwoEpochAgo == nil || err != nil {
-	// 	utils.Logger().Debug().
-	// 		Msgf("apr compute headers epochs ago %+v %+v %+v %+v %+v %+v",
-	// 			twoEpochAgo, oneEpochAgo,
-	// 			blockNumAtTwoEpochAgo, blockNumAtOneEpochAgo,
-	// 			headerOneEpochAgo, headerTwoEpochAgo,
-	// 		)
-	// 	return &zero, nil
-	// }
-
-	// utils.Logger().Info().
-	// 	RawJSON("current-epoch-header", []byte(bc.CurrentHeader().String())).
-	// 	RawJSON("one-epoch-ago-header", []byte(headerOneEpochAgo.String())).
-	// 	RawJSON("two-epoch-ago-header", []byte(headerTwoEpochAgo.String())).
-	// 	Msg("headers used for apr computation")
-
-	// estimatedRewardPerYear, err := expectedRewardPerYear(
-	// 	headerOneEpochAgo, headerTwoEpochAgo,
-	// 	oneSnapshotAgo, twoSnapshotAgo,
-	// 	blocksPerEpoch,
-	// )
-
-	// if err != nil {
-	// 	return nil, err
-	// }
-
-	// if estimatedRewardPerYear.Cmp(common.Big0) == 0 {
-	// 	return &zero, nil
-	// }
-
-	// total := numeric.NewDecFromBigInt(validatorNow.TotalDelegation())
-	// if total.IsZero() {
-	// 	return nil, errors.New("zero total delegation will cause div by zero")
-	// }
-
-	// result := numeric.NewDecFromBigInt(estimatedRewardPerYear).Quo(
-	// 	total,
-	// )
-	// return &result, nil
+	oneEpochAgo, zero :=
+		new(big.Int).Sub(block.Epoch(), common.Big1),
+		numeric.ZeroDec()
+
+	utils.Logger().Info().
+		Uint64("now", block.Epoch().Uint64()).
+		Uint64("one-epoch-ago", oneEpochAgo.Uint64()).
+		Msg("apr - begin compute for validator ")
+
+	oneSnapshotAgo, err := bc.ReadValidatorSnapshotAtEpoch(
+		oneEpochAgo,
+		validatorNow.Address,
+	)
+
+	if err != nil {
+		return &zero, nil
+	}
+
+	blockNumAtOneEpochAgo := shard.Schedule.EpochLastBlock(oneEpochAgo.Uint64())
+
+	headerOneEpochAgo := bc.GetHeaderByNumber(blockNumAtOneEpochAgo)
+	if block.Header() == nil || headerOneEpochAgo == nil || err != nil {
+		utils.Logger().Debug().
+			Msgf("apr compute headers epochs ago %+v %+v %+v",
+				oneEpochAgo,
+				blockNumAtOneEpochAgo,
+				headerOneEpochAgo,
+			)
+		return &zero, nil
+	}
+
+	utils.Logger().Info().
+		RawJSON("current-epoch-header", []byte(bc.CurrentHeader().String())).
+		RawJSON("one-epoch-ago-header", []byte(headerOneEpochAgo.String())).
+		Msg("headers used for apr computation")
+
+	estimatedRewardPerYear, err := expectedRewardPerYear(
+		block.Header(), headerOneEpochAgo,
+		validatorNow, oneSnapshotAgo,
+	)
+
+	if err != nil {
+		return nil, err
```

Comment:
why return `nil` over `&zero` ? I am bit confused about the return value specification.

Reply:
fixed. the errors should be returned.